src/test/fixtures/mocks/analytics.js

```diff
@@ -0,0 +1,23 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import type { GoogleAnalytics, GAPayload } from '../../../utils/analytics';
+
+/**
+ * This function records all of the analytics events that are fired through side effects,
+ * but then handles all cleanup after running the callback function that actually
+ * holds the test.
+ */
+export function withAnalyticsMock<F: (Array<[string, GAPayload]>) => void>(
+  callback: F
+) {
+  const analyticsEvents = [];
+  const ga: GoogleAnalytics = (type: string, payload: *) => {
+    analyticsEvents.push([type, payload]);
+  };
+  self.ga = ga;
+  callback(analyticsEvents);
+  delete self.ga;
+}
```

Comment:
I won't ask you to change this now. But you could have used `jest.mock` or `sinon.stub`. Then the assertions could have used `self.ga.calls` (for jest) or `sinon.assert.calledWith` (for sinon) directly.

What's disturbs me most is the fact that this works because `analyticsEvents` is shared between the `ga` function here and the callback. Could you please add a comment to state this?

Reply:
Thanks for pushing on this, I switched it the `jest.fn()`. This feels more idiomatic, while still containing side-effects and making sure cleanup happens.